src/resources/dart-3-migration.md

```diff
@@ -0,0 +1,512 @@
+resolution even when the upper bound is limited to versions below 3.0.0. For
+example, a package with the following constraint will be allowed to resolve with
+a Dart 3.x SDK, as pub will re-interpret the upper-constraint `<3.0.0` as `<4.0.0`
+when the lower constraint is `2.12` or higher:
+
+```yaml
+environment:
+  sdk: '>=2.14.0 <3.0.0'           # This is interpreted as '>=2.14.0 <4.0.0'
+```
+
+This allows developers to use Dart 3 sound null safety with packages that have
+been migrated to 2.12 null safety without needing a second migration, unless
+the code is affected by any other Dart 3 changes.
+
+### Testing for impact
+
+To understand if your source code is impacted by any Dart 3 changes, 
+use these steps:
+
+```terminal
+$ dart --version    # make sure this reports 3.0.0 or higher
+$ dart pub get      # this should resolve without issues
+$ dart analyze      # this should pass without errors
+```
+
+If the `pub get` step fails, try to upgrade your dependencies
+to see if more recent versions might support Dart 3:
+
+```terminal
+$ dart pub upgrade
+$ dart analyze      # this should pass without errors
+```
+
+Or if needed also include major versions upgrades:
+```terminal
+$ dart pub upgrade --major-versions
+$ dart analyze      # this should pass without errors
+```
+
+## Dart 3 language changes
+
+### 100% sound null safety
+
+Null safety was introduced in Dart 2.12, more than two years ago. In Dart 2.12
+it was enabled [with a pubspec setting](/null-safety/#enable-null-safety),
+but in Dart 3, you always get null safety, and it cannot be turned off.
+
+#### Scope
+
+This is an *unversioned* change, that applies to all Dart 3 code.
+
+#### Symptom
+
+Packages developed without null safety support will cause issues
+when resolving dependencies with `pub get`:
+
+```terminal
+$ dart pub get
+
+Because pkg1 doesn't support null safety, version solving failed.
+The lower bound of "sdk: '>=2.9.0 <3.0.0'" must be 2.12.0 or higher to enable null safety.
+```
+
+Libraries that opt out of null safety with [language version comments](/guides/language/evolution#per-library-language-version-selection)
+that select any language version below `2.12` will cause analysis or compilation
+errors:
+
+```terminal
+$ dart analyze .
+Analyzing ....                         0.6s
+
+  error • lib/pkg1.dart:1:1 • The language version must be >=2.12.0. 
+  Try removing the language version override and migrating the code.
+  • illegal_language_version_override
+```
+
+```terminal
+$ dart run bin/my_app.dart
+../pkg1/lib/pkg1.dart:1:1: Error: Library doesn't support null safety.
+// @dart=2.9
+^^^^^^^^^^^^
+```
+
+#### Migration
+
+A first step before beginning any migration to Dart 3, is to ensure your app or package has been 100% migrated to enable null safety. This must happen using a Dart `2.19` SDK, not a Dart 3 SDK. For details, see the [null safety migration guide](/null-safety/migration-guide).
+
+### Colon-syntax for default values
```

Comment:
Nothing to change here, but this needs to be updated in the "named parameters" section of the [Functions](https://dart.dev/language/functions#named-parameters) page. Will make an issue for tracking